### PR TITLE
feat: stage assets under .cdk.assets

### LINKED ISSUE #1716: JSII libraries with assets not found
When trying to use a JSII-built library with assets, those assets will not be found at deployment time. For example, if the library has embedded lambda code, the metadata for the asset will point to a temporary folder. This may be the reason why the .zip file in S3 is empty.

Interesting line in synth: ```aws:asset:path: /private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-tL6ToE/node_modules/cdk-dynamo-table-viewer/lib/lambda```

Library used: cdk-dynamo-table-viewer
Binding Language Used: C# (.NET)
Synth output:
```yaml
ViewHitCounterRendered9C783E45:
    Type: AWS::Lambda::Function
    Properties:
      Code:
        S3Bucket:
          Ref: ViewHitCounterRenderedCodeS3Bucket7AAA6C2C
        S3Key:
          Fn::Join:
            - ""
            - - Fn::Select:
                  - 0
                  - Fn::Split:
                      - "||"
                      - Ref: ViewHitCounterRenderedCodeS3VersionKey7A5EC4B1
              - Fn::Select:
                  - 1
                  - Fn::Split:
                      - "||"
                      - Ref: ViewHitCounterRenderedCodeS3VersionKey7A5EC4B1
      Handler: index.handler
      Role:
        Fn::GetAtt:
          - ViewHitCounterRenderedServiceRole254DB4EA
          - Arn
      Runtime: nodejs8.10
      Environment:
        Variables:
          TABLE_NAME:
            Ref: HelloHitCounterHits7AAEBF80
          TITLE: Hello Hits
          SORT_BY: -hits
    DependsOn:
      - ViewHitCounterRenderedServiceRoleDefaultPolicy9ADB8C83
      - ViewHitCounterRenderedServiceRole254DB4EA
    Metadata:
      aws:cdk:path: CdkWorkshopStack/ViewHitCounter/Rendered/Resource
      aws:asset:path: /private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-tL6ToE/node_modules/cdk-dynamo-table-viewer/lib/lambda
      aws:asset:property: Code
```
deploy trace:
```sh
 - /private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-4sE9J8/node_modules/cdk-dynamo-table-viewer/lib/lambda (zip)
Preparing asset {"path":"/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-4sE9J8/node_modules/cdk-dynamo-table-viewer/lib/lambda","id":"CdkWorkshopStackViewHitCounterRenderedCode7E728AF9","packaging":"zip","s3BucketParameter":"ViewHitCounterRenderedCodeS3Bucket7AAA6C2C","s3KeyParameter":"ViewHitCounterRenderedCodeS3VersionKey7A5EC4B1"}
Preparing zip asset from directory: /private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-4sE9J8/node_modules/cdk-dynamo-table-viewer/lib/lambda
zip archive: /var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/cdk-assetsjmwzs1/archive.zip
Preparing file asset: /var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/cdk-assetsjmwzs1/archive.zip
s3://cdktoolkit-stagingbucket----------/assets/CdkWorkshopStackViewHitCounterRenderedCode7E728AF9/8739c76e681f900923b900c9df0ef75cf421d39cabb54650c4b9ad19b6a76d85.zip: checking if already exists
s3://cdktoolkit-stagingbucket----------/assets/CdkWorkshopStackViewHitCounterRenderedCode7E728AF9/8739c76e681f900923b900c9df0ef75cf421d39cabb54650c4b9ad19b6a76d85.zip: uploading
s3://cdktoolkit-stagingbucket---------/assets/CdkWorkshopStackViewHitCounterRenderedCode7E728AF9/8739c76e681f900923b900c9df0ef75cf421d39cabb54650c4b9ad19b6a76d85.zip: upload complete
S3 url for ../../../../private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-4sE9J8/node_modules/cdk-dynamo-table-viewer/lib/lambda: s3://cdktoolkit-stagingbucket--------/assets/CdkWorkshopStackViewHitCounterRenderedCode7E728AF9/8739c76e681f900923b900c9df0ef75cf421d39cabb54650c4b9ad19b6a76d85.zip
Updated: ../../../../private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-4sE9J8/node_modules/cdk-dynamo-table-viewer/lib/lambda (zip)
s3://cdktoolkit-stagingbucket------
```

```sh
15/37 | 11:12:07 PM | CREATE_FAILED        | AWS::Lambda::Function       | ViewHitCounter/Rendered (ViewHitCounterRendered9C783E45) Uploaded file must be a non-empty zip (Service: AWSLambdaInternal; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: ----------------)
        new Function (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-4sE9J8/node_modules/@aws-cdk/aws-lambda/lib/function.js:62:26)
        \_ new TableViewer (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/jsii-kernel-4sE9J8/node_modules/cdk-dynamo-table-viewer/lib/table-viewer.js:14:25)
        \_ s._wrapSandboxCode (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:86829)
        \_ t.Kernel._wrapSandboxCode (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:95312)
        \_ t.Kernel._create (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:86808)
        \_ t.Kernel.create (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:81792)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78503)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ t.KernelHost.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78957)
        \_ t.KernelHost.run (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78192)
        \_ e.processRequest (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78218)
        \_ i.then.e (/private/var/folders/0t/_hd2htfd497g3b2z41wlvlx439_1m_/T/fuxkshgm.mys/jsii-runtime.js:1:78831)
```

### LINKED ISSUE #2096: AWS::Lambda::Function should store relative path instead of absolute path
After checking out repo on different machine, `cdk diff` shows

```
Resources
[~] AWS::Lambda::Function AdoptEcrRepositorydbc60defc59544bcaa5c28c95d68f62c AdoptEcrRepositorydbc60defc59544bcaa5c28c95d68f62c52BE89E9 
 └─ [~] Metadata
     └─ [~] .aws:asset:path:
         ├─ [-] /Users/apple/GitProject/hello-cdk/node_modules/@aws-cdk/assets-docker/lib/adopt-repository
         └─ [+] /Users/apple/GitProject/ECS_Cluster_CDK/node_modules/@aws-cdk/assets-docker/lib/adopt-repository
```

which cause unnecessary deployment.

### PR DESCRIPTION
**Pull request by eladb**
To ensure that assets are available for the toolchain to deploy after the CDK app
exists, the CLI will, by default, request that the app will stage the assets under
the `.cdk.assets` directory (relative to working directory).

The CDK will then *copy* all assets from their source locations to this staging
directory and will refer to the staging location as the asset path. Assets will
be stored using their content fingerprint (md5 hash) so they will never be copied
twice unless they change.

Docker build context directories will also be staged.

Staging is disabled by default and in cdk-integ.

Added .cdk.staging to all .gitignore files in cdk init templates.

Fixes #1716
Fixes #2096

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by rix0rrr** [comment]
Does that mean the tempdir will go in the user's app directory?

**Comment by eladb** [comment]
> I don't care for the default location of the staging directory. I think it's going to invite more questions and discussions. But it'll be easy enough to change, so okay.

Can you elaborate what you don’t care about?



> By the way, this PR could be trivially changed to also fix #2096, by having assets.Staging output the relative file path and the assets handers in the toolkit resolving directories relative to the staging root.

It actually fixes this already. Added.

**Comment by rix0rrr** [comment]
> Can you elaborate what you don’t care about?

First, everyone who has a CDK project today will not have this directory in their `.gitignore` yet, and this change will start generating more files in their source directory, which a careless `git add -A` is going to commit to version control.

Second, the directory is not going to be as inconspicuous as you think on Windows. 

I'd prefer for this directory to be off to the side somewhere, maybe in $TMP

**Comment by eladb** [comment]
> I'd prefer for this directory to be off to the side somewhere, maybe in $TMP

How about $HOME? This way, assets are reused across CDK apps.

**Comment by rix0rrr** [comment]
> This way, assets are reused across CDK apps.

If we use a stable name under $TMP we can also reuse across CDK apps, and I think putting there covers the intent better. Or it would have to be `$HOME/.cache`, which I think is also an (emerging?) standard, but doesn't place so nicely on Windows. Windows would have to be `%LOCALAPPDATA%` if we wanted to be correct.

**Comment by eladb** [comment]
We use stable names (based off of the content fingerprint). It is a short of cache.

I am leaning towards leaving this under the project directory though as it aligns better with our longer term Toolchain project where “synth” and “bundle” happen on different machines so we need to move all the asset sources between those machines.

**Comment by rix0rrr** [comment]
In that case, are the default settings maybe optimizing for CI builds, thereby accidentally pessimizing local builds?

If so, we could switch the default off of the `--ci` flag.

File: packages/@aws-cdk/assets/lib/fs/fingerprint.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    hash.update(`==exclude==${ex}==\n\n`);
+  }
+
+  return hash.digest('hex');
+
+  function addToHash(pathToAdd: string) {
+    hash.update('==\n');
+    const relativePath = path.relative(fileOrDirectory, pathToAdd);
+    hash.update(relativePath + '\n');
+    hash.update('~~~~~~~~~~~~~~~~~~\n');
```
Comment (sam-goodwin) [inline packages/@aws-cdk/assets/lib/fs/fingerprint.ts]: What's with the arbitrary strings?

Comment (eladb) [reply packages/@aws-cdk/assets/lib/fs/fingerprint.ts]: Just separators between the various components of the hash so they don't conflict.

**Comment by eladb** [comment]
@rix0rrr decided for now to leave it under the project directory. @sam-goodwin is going to refactor this quite a bit in his work on the toolchain and I rather keep this simple for now.

**Pull request merged**